torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -247,6 +247,76 @@ static PyObject* applyBoolGetitem(const Variable& self, bool index) {
   }
 }
 
+static bool canDispatchToLegacyIndexing(const Variable& self, const variable_list& vars) {
+  // mask indexing
+  if (vars.size() == 1 && vars[0].type().scalarType() == ScalarType::Byte && vars[0].is_same_size(self)) {
+    return true;
+  }
+
+  // single tensor indexing
+  int num_defined_variables = 0;
+  for (auto& variable : vars) {
+    auto is_defined = variable.defined();
+    num_defined_variables += is_defined;
+    if (is_defined && (variable.dim() != 1 || variable.type().scalarType() != ScalarType::Long || variable.numel() == 0)) {
+      num_defined_variables = -1;
+      break;
+    }
+  }
+  if (num_defined_variables == 1) {
+    return true;
+  }
+  // advanced indexing
+  return false;
+}
+
+static Variable dispatch_legacy_index(const Variable& self, const variable_list& vars) {
+  if (vars.size() == 1 && vars[0].type().scalarType() == ScalarType::Byte && vars[0].is_same_size(self)) {
```

Comment:
Instead of repeating these checks exactly I think it would be cleaner if canDispatchToLegacyIndexing was renamed legacyIndexingDispatchType and returned an enum in {None, Mask, Select}.  Then you just check those here and don't need the checks.

Reply:
You could also return a pair which includes the relevant dimension if the type isn't None.

Reply:
I agree, and that was my first thought as well, but I decided to remove it for simplicity. But overall I think it would be better to avoid repeating those checks